diff -ru papi-5.6.1.0-orig/src/validation_tests/cycles_validation.c papi-5.6.1.0/src/validation_tests/cycles_validation.c
--- papi-5.6.1.0-orig/src/validation_tests/cycles_validation.c	2018-09-22 04:26:30.000000000 -0600
+++ papi-5.6.1.0/src/validation_tests/cycles_validation.c	2018-09-22 04:27:20.000000000 -0600
@@ -151,7 +151,7 @@
 	}
 
 	/* Check that TOT_INS is reasonable */
-	if (abs(values[1] - (1000000*NUM_LOOPS)) > (1000000*NUM_LOOPS)) {
+	if (llabs(values[1] - (1000000*NUM_LOOPS)) > (1000000*NUM_LOOPS)) {
 		printf("%s Error of %.2f%%\n", "PAPI_TOT_INS", (100.0 * (double)(values[1] - (1000000*NUM_LOOPS)))/(1000000*NUM_LOOPS));
 		test_fail( __FILE__, __LINE__, "Instruction validation", 0 );
 	}
diff -ru papi-5.6.1.0-orig/src/validation_tests/instructions_testcode.c papi-5.6.1.0/src/validation_tests/instructions_testcode.c
--- papi-5.6.1.0-orig/src/validation_tests/instructions_testcode.c	2018-09-22 04:26:30.000000000 -0600
+++ papi-5.6.1.0/src/validation_tests/instructions_testcode.c	2018-09-22 04:28:56.000000000 -0600
@@ -77,7 +77,7 @@
 		"	bne	test_loop	// repeat till zero\n"
 		: /* no output registers */
 		: /* no inputs */
-		: "cc", "r2" /* clobbered */
+		: "cc", "x2" /* clobbered */
 	);
 	return 0;
 #endif
